File: src/_2015/Permutation.java
// September 1st, 2020 - Used in 9a, 9b, 13a, and 13b
package _2015;

import java.util.ArrayList;
import java.util.List;

public class Permutation {

    public static <T> List<List<T>> create(List<T> list) {

        List<List<T>> permutations = new ArrayList<List<T>>();
        permutations.add(new ArrayList<T>());
        for (T item : list) {
            permutations = add(item, permutations);
        }
        return permutations;

    }

    private static <T> List<List<T>> add(T item, List<List<T>> list) {
        List<List<T>> permutations = new ArrayList<List<T>>();
        for (List<T> permutation : list) {
            for (int i = 0; i < permutation.size() + 1; i++) {
                List<T> copy = new ArrayList<T>(permutation);
                copy.add(i, item);
                permutations.add(copy);
            }
        }
        return permutations;
    }

}

File: Tool/JavaScript/Permutation.js
class Permutation {

	static create(...list) {
		let permutations = [[list[0]]];
		for (var i = 1; i < list.length; i++) {
			permutations = Permutation.add(permutations, list[i]);
		}
		return permutations;
	}

	static add(permutations, item) {
		let temp = [];
		for (var perm of permutations) {
			for (var j = 0; j <= perm.length; j++) {
				let copy = perm.slice();
				copy.splice(j, 0, item);
				temp.push(copy);
			}
		}
		return temp;
	}

}

module.exports = Permutation;
